src/QsCompiler/Transformations/CallGraph.cs

```diff
@@ -68,64 +160,144 @@ internal static bool TryCombineTypeResolutionsForTarget(QsQualifiedName target,
                 {
-                    // Get all the parameters whose value is dependent on the current resolution's type parameter,
-                    // and update their values with this resolution's value.
-                    foreach (var keyInCombined in mayBeReplaced[typeParam])
-                    {
-                        // Check that we are not constricting a type parameter to another type parameter of the same callable.
-                        success = success && !IsConstrictiveResolution(keyInCombined, paramRes);
-                        combinedBuilder[keyInCombined] = paramRes;
-                    }
+                    success = UpdatedReplaceableResolutions(success, mayBeReplaced, combinedBuilder, typeParam, paramRes);
                 }
 
-                // Add all resolutions to the current dictionary.
-                foreach (var (typeParam, paramRes) in resolutionDictionary)
+                // Validate and add each resolution to the result.
+                foreach (var (typeParam, paramRes) in resolvedDictionary)
                 {
                     // Check that we are not constricting a type parameter to another type parameter of the same callable.
-                    success = success && !IsConstrictiveResolution(typeParam, paramRes);
+                    success = success && !ConstrictionCheck.Apply(typeParam, paramRes);
+
                     // Check that there is no conflicting resolution already defined.
                     var conflictingResolutionExists = combinedBuilder.TryGetValue(typeParam, out var current)
                         && !current.Equals(paramRes) && !IsSelfResolution(typeParam, current);
                     success = success && !conflictingResolutionExists;
+
+                    // Add the resolution to the current dictionary.
                     combinedBuilder[typeParam] = paramRes;
                 }
             }
 
-            combined = combinedBuilder.ToImmutable();
+            combinedResolutions = combinedBuilder.ToImmutable();
             return success;
         }
+
+        /// <summary>
+        /// Walker that collects all of the type parameter references for a given ResolvedType
+        /// and returns them as a HashSet.
+        /// </summary>
+        internal class GetTypeParameters : TypeTransformation<GetTypeParameters.TransformationState>
+        {
+            /// <summary>
+            /// Walks the given ResolvedType and returns all of the type parameters referenced.
+            /// </summary>
+            public static HashSet<TypeParameterName> Apply(ResolvedType res)
+            {
+                var walker = new GetTypeParameters();
+                walker.OnType(res);
+                return walker.SharedState.TypeParams;
+            }
+
+            internal class TransformationState
+            {
+                public HashSet<TypeParameterName> TypeParams = new HashSet<TypeParameterName>();
```

Comment:
You could directly make this transformation a `TypeTransformation<HashSet<TypeParameterName>>`.

Reply:
I could skip over having an explicit TransformationState class, but then it is not as clear what is being used when referencing `this.SharedState` in the code of the transformation. `this.SharedState.TypeParams.Add(...)` is more clear to me than `this.SharedState.Add(...)`. In general I prefer to have a wrapper class for my state when writing and dealing with transformations.